File: security/security-core/src/main/java/io/camunda/security/auth/Authorization.java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */
package io.camunda.security.auth;

import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.AUDIT_LOG;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.AUTHORIZATION;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.BATCH;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.CLUSTER_VARIABLE;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.DECISION_DEFINITION;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.DECISION_REQUIREMENTS_DEFINITION;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.DOCUMENT;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.GLOBAL_LISTENER;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.GROUP;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.MAPPING_RULE;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.PROCESS_DEFINITION;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.ROLE;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.SYSTEM;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.TENANT;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.USER;
import static io.camunda.zeebe.protocol.record.value.AuthorizationResourceType.USER_TASK;
import static io.camunda.zeebe.protocol.record.value.AuthorizationScope.WILDCARD;
import static io.camunda.zeebe.protocol.record.value.PermissionType.CREATE_PROCESS_INSTANCE;
import static io.camunda.zeebe.protocol.record.value.PermissionType.DELETE_DECISION_INSTANCE;
import static io.camunda.zeebe.protocol.record.value.PermissionType.DELETE_PROCESS_INSTANCE;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_DECISION_DEFINITION;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_DECISION_INSTANCE;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_JOB_METRIC;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_PROCESS_DEFINITION;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_PROCESS_INSTANCE;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_USAGE_METRIC;
import static io.camunda.zeebe.protocol.record.value.PermissionType.READ_USER_TASK;
import static io.camunda.zeebe.protocol.record.value.PermissionType.UPDATE_PROCESS_INSTANCE;
import static io.camunda.zeebe.protocol.record.value.PermissionType.UPDATE_USER_TASK;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.camunda.zeebe.protocol.record.value.AuthorizationResourceType;
import io.camunda.zeebe.protocol.record.value.PermissionType;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Predicate;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public record Authorization<T>(
    @JsonProperty("resource_type") AuthorizationResourceType resourceType,
    @JsonProperty("permission_type") PermissionType permissionType,
    @JsonProperty("resource_ids") List<String> resourceIds,
    @JsonIgnore Function<T, String> resourceIdSupplier,
    @JsonProperty("resource_property_names") Set<String> resourcePropertyNames,
    @JsonIgnore Predicate<T> condition,
    @JsonProperty("transitive") boolean transitive) {

  // USER TASK property names
  public static final String PROP_ASSIGNEE = "assignee";
  public static final String PROP_CANDIDATE_USERS = "candidateUsers";
  public static final String PROP_CANDIDATE_GROUPS = "candidateGroups";

  public boolean hasAnyResourceIds() {
    return resourceIds != null && !resourceIds.isEmpty();
  }

  public boolean hasAnyResourcePropertyNames() {
    return resourcePropertyNames != null && !resourcePropertyNames.isEmpty();
  }

  public boolean hasAnyResourceAccess() {
    return hasAnyResourceIds() || hasAnyResourcePropertyNames();
  }

  public static <T> Authorization<T> withAuthorization(
      final Authorization<T> authorization, final String resourceId) {
    return authorization.withResourceId(resourceId);
  }

  public static <T> Authorization<T> withAuthorization(
      final Authorization<T> authorization, final Function<T, String> resourceIdSupplier) {
    return authorization.withResourceIdSupplier(resourceIdSupplier);
  }

  public Authorization<T> withResourceId(final String resourceId) {
    return withResourceIds(List.of(resourceId));
  }

  public Authorization<T> withResourceIds(final List<String> resourceIds) {
    return new Authorization<>(
        resourceType(),
        permissionType(),
        List.copyOf(resourceIds),
        resourceIdSupplier(),
        resourcePropertyNames(),
        condition(),
        transitive());
  }

  public Authorization<T> withResourceIdSupplier(final Function<T, String> resourceIdSupplier) {
    return new Authorization<>(
        resourceType(),
        permissionType(),
        resourceIds(),
        resourceIdSupplier,
        resourcePropertyNames(),
        condition(),
        transitive());
  }

  public Authorization<T> withResourcePropertyNames(final Set<String> resourcePropertyNames) {
    return new Authorization<>(
        resourceType(),
        permissionType(),
        resourceIds(),
        resourceIdSupplier(),
        resourcePropertyNames,
        condition(),
        transitive());
  }

  public Authorization<T> withCondition(final Predicate<T> condition) {
    return new Authorization<>(
        resourceType(),
        permissionType(),
        resourceIds(),
        resourceIdSupplier(),
        resourcePropertyNames(),
        condition,
        transitive());
  }

  public static <T> Authorization<T> of(final Function<Builder<T>, Builder<T>> builderFunction) {
    return builderFunction.apply(new Builder<>()).build();
  }

  public boolean appliesTo(final T document) {
    return condition == null || condition.test(document);
  }

  @JsonIgnore
  public boolean isWildcard() {
    return resourceIds != null
        && resourceIds.stream().anyMatch(id -> WILDCARD.getResourceId().equals(id));
  }

  public static class Builder<T> {
    private AuthorizationResourceType resourceType;
    private PermissionType permissionType;
    private List<String> resourceIds;
    private Function<T, String> resourceIdSupplier;
    private Set<String> resourcePropertyNames;
    private Predicate<T> condition = null;
    private boolean transitive = false;

    public Builder<T> resourceType(final AuthorizationResourceType resourceType) {
      this.resourceType = resourceType;
      return this;
    }

    public Builder<T> permissionType(final PermissionType permissionType) {
      this.permissionType = permissionType;
      return this;
    }

    public Builder<T> condition(final Predicate<T> condition) {
      this.condition = condition;
      return this;
    }

    public Builder<T> transitive() {
      transitive = true;
      return this;
    }

    public Builder<T> processDefinition() {
      return resourceType(PROCESS_DEFINITION);
    }

    public Builder<T> decisionDefinition() {
      return resourceType(DECISION_DEFINITION);
    }

    public Builder<T> decisionRequirementsDefinition() {
      return resourceType(DECISION_REQUIREMENTS_DEFINITION);
    }

    public Builder<T> mappingRule() {
      return resourceType(MAPPING_RULE);
    }

    public Builder<T> role() {
      return resourceType(ROLE);
    }

    public Builder<T> group() {
      return resourceType(GROUP);
    }

    public Builder<T> tenant() {
      return resourceType(TENANT);
    }

    public Builder<T> authorization() {
      return resourceType(AUTHORIZATION);
    }

    public Builder<T> user() {
      return resourceType(USER);
    }

    public Builder<T> userTask() {
      return resourceType(USER_TASK);
    }

    public Builder<T> system() {
      return resourceType(SYSTEM);
    }

    public Builder<T> read() {
      return permissionType(READ);
    }

    public Builder<T> readProcessDefinition() {
      return permissionType(READ_PROCESS_DEFINITION);
    }

    public Builder<T> readDecisionDefinition() {
      return permissionType(READ_DECISION_DEFINITION);
    }

    public Builder<T> readProcessInstance() {
      return permissionType(READ_PROCESS_INSTANCE);
    }

    public Builder<T> createProcessInstance() {
      return permissionType(CREATE_PROCESS_INSTANCE);
    }

    public Builder<T> updateProcessInstance() {
      return permissionType(UPDATE_PROCESS_INSTANCE);
    }

    public Builder<T> readUserTask() {
      return permissionType(READ_USER_TASK);
    }

    public Builder<T> updateUserTask() {
      return permissionType(UPDATE_USER_TASK);
    }

    public Builder<T> readDecisionInstance() {
      return permissionType(READ_DECISION_INSTANCE);
    }

    public Builder<T> readUsageMetric() {
      return permissionType(READ_USAGE_METRIC);
    }

    public Builder<T> readJobMetric() {
      return permissionType(READ_JOB_METRIC);
    }

    public Builder<T> deleteProcessInstance() {
      return permissionType(DELETE_PROCESS_INSTANCE);
    }

    public Builder<T> deleteDecisionInstance() {
      return permissionType(DELETE_DECISION_INSTANCE);
    }

    public Builder<T> batchOperation() {
      return resourceType(BATCH);
    }

    public Builder<T> document() {
      return resourceType(DOCUMENT);
    }

    public Builder<T> auditLog() {
      return resourceType(AUDIT_LOG);
    }

    public Builder<T> clusterVariable() {
      return resourceType(CLUSTER_VARIABLE);
    }

    public Builder<T> globalListener() {
      return resourceType(GLOBAL_LISTENER);
    }

    public Builder<T> resourceId(final String resourceId) {
      return resourceIds(List.of(resourceId));
    }

    public Builder<T> resourceIds(final List<String> resourceIds) {
      this.resourceIds = resourceIds;
      return this;
    }

    public Builder<T> resourceIdSupplier(final Function<T, String> resourceIdSupplier) {
      this.resourceIdSupplier = resourceIdSupplier;
      return this;
    }

    public Builder<T> resourcePropertyNames(final Set<String> resourcePropertyNames) {
      this.resourcePropertyNames = resourcePropertyNames;
      return this;
    }

    public Builder<T> authorizedByProperty(final String propertyName) {
      if (this.resourcePropertyNames == null) {
        this.resourcePropertyNames = new HashSet<>();
      }
      this.resourcePropertyNames.add(propertyName);
      return this;
    }

    public Builder<T> authorizedByAssignee() {
      return authorizedByProperty(PROP_ASSIGNEE);
    }

    public Builder<T> authorizedByCandidateUsers() {
      return authorizedByProperty(PROP_CANDIDATE_USERS);
    }

    public Builder<T> authorizedByCandidateGroups() {
      return authorizedByProperty(PROP_CANDIDATE_GROUPS);
    }

    /**
     * Fluent connector indicating the next condition is an OR alternative. This is purely for
     * readability.
     */
    public Builder<T> or() {
      return this;
    }

    public Authorization<T> build() {
      return new Authorization<>(
          resourceType,
          permissionType,
          resourceIds,
          resourceIdSupplier,
          resourcePropertyNames,
          condition,
          transitive);
    }
  }
}

File: webapp/client/packages/camunda-api-zod-schemas/lib/8.10/authorization.ts
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {z} from 'zod';
import {API_VERSION, getQueryResponseBodySchema, getQueryRequestBodySchema, type Endpoint} from './common';

const permissionTypeSchema = z.enum([
	'ACCESS',
	'CANCEL_PROCESS_INSTANCE',
	'CLAIM',
	'COMPLETE',
	'CREATE',
	'CREATE_BATCH_OPERATION_CANCEL_PROCESS_INSTANCE',
	'CREATE_BATCH_OPERATION_DELETE_PROCESS_INSTANCE',
	'CREATE_BATCH_OPERATION_MIGRATE_PROCESS_INSTANCE',
	'CREATE_BATCH_OPERATION_MODIFY_PROCESS_INSTANCE',
	'CREATE_BATCH_OPERATION_RESOLVE_INCIDENT',
	'CREATE_BATCH_OPERATION_DELETE_DECISION_INSTANCE',
	'CREATE_BATCH_OPERATION_DELETE_DECISION_DEFINITION',
	'CREATE_BATCH_OPERATION_DELETE_PROCESS_DEFINITION',
	'CREATE_BATCH_OPERATION_SUSPEND_PROCESS_INSTANCE',
	'CREATE_PROCESS_INSTANCE',
	'CREATE_DECISION_INSTANCE',
	'CREATE_TASK_LISTENER',
	'READ',
	'READ_JOB_METRIC',
	'READ_PROCESS_INSTANCE',
	'READ_TASK_LISTENER',
	'READ_USAGE_METRIC',
	'READ_USER_TASK',
	'READ_DECISION_INSTANCE',
	'READ_PROCESS_DEFINITION',
	'READ_DECISION_DEFINITION',
	'UPDATE',
	'UPDATE_TASK_LISTENER',
	'UPDATE_PROCESS_INSTANCE',
	'UPDATE_USER_TASK',
	'DELETE',
	'DELETE_PROCESS',
	'DELETE_DRD',
	'DELETE_FORM',
	'DELETE_RESOURCE',
	'DELETE_PROCESS_INSTANCE',
	'DELETE_DECISION_INSTANCE',
	'DELETE_TASK_LISTENER',
	'EVALUATE',
	'MODIFY_PROCESS_INSTANCE',
	'SUSPEND_PROCESS_INSTANCE',
	'CLAIM_USER_TASK',
	'COMPLETE_USER_TASK',
	'REVEAL',
]);
type PermissionType = z.infer<typeof permissionTypeSchema>;

const resourceTypeSchema = z.enum([
	'AUDIT_LOG',
	'AUTHORIZATION',
	'BATCH',
	'CLUSTER_VARIABLE',
	'COMPONENT',
	'DECISION_DEFINITION',
	'DECISION_REQUIREMENTS_DEFINITION',
	'DOCUMENT',
	'EXPRESSION',
	'GLOBAL_LISTENER',
	'GROUP',
	'MAPPING_RULE',
	'MESSAGE',
	'PROCESS_DEFINITION',
	'RESOURCE',
	'ROLE',
	'SECRET',
	'SYSTEM',
	'TENANT',
	'USER',
	'USER_TASK',
]);
type ResourceType = z.infer<typeof resourceTypeSchema>;

const ownerTypeSchema = z.enum(['USER', 'CLIENT', 'ROLE', 'GROUP', 'MAPPING_RULE', 'UNSPECIFIED']);
type OwnerType = z.infer<typeof ownerTypeSchema>;

const authorizationSchema = z.object({
	ownerId: z.string(),
	ownerType: ownerTypeSchema,
	resourceType: resourceTypeSchema,
	resourceId: z.string().nullable(),
	resourcePropertyName: z.string().nullable(),
	permissionTypes: z.array(permissionTypeSchema),
	authorizationKey: z.string(),
});
type Authorization = z.infer<typeof authorizationSchema>;

const createAuthorizationRequestBodySchema = authorizationSchema.pick({
	ownerId: true,
	ownerType: true,
	resourceType: true,
	resourceId: true,
	permissionTypes: true,
});
type CreateAuthorizationRequestBody = z.infer<typeof createAuthorizationRequestBodySchema>;

const createAuthorizationResponseBodySchema = z.object({
	authorizationKey: z.string(),
});
type CreateAuthorizationResponseBody = z.infer<typeof createAuthorizationResponseBodySchema>;

const updateAuthorizationRequestBodySchema = createAuthorizationRequestBodySchema.partial();
type UpdateAuthorizationRequestBody = z.infer<typeof updateAuthorizationRequestBodySchema>;

const queryAuthorizationsRequestBodySchema = getQueryRequestBodySchema({
	sortFields: ['ownerId', 'ownerType', 'resourceId', 'resourceType'] as const,
	filter: z
		.object({
			resourceId: z.array(z.string()),
			...authorizationSchema.pick({
				ownerId: true,
				ownerType: true,
				resourceType: true,
			}).shape,
		})
		.partial(),
});
type QueryAuthorizationsRequestBody = z.infer<typeof queryAuthorizationsRequestBodySchema>;

const getAuthorizationResponseBodySchema = authorizationSchema;
type GetAuthorizationResponseBody = z.infer<typeof getAuthorizationResponseBodySchema>;

const queryAuthorizationsResponseBodySchema = getQueryResponseBodySchema(authorizationSchema);
type QueryAuthorizationsResponseBody = z.infer<typeof queryAuthorizationsResponseBodySchema>;

const createAuthorization = {
	method: 'POST',
	getUrl: () => `/${API_VERSION}/authorizations` as const,
} as const satisfies Endpoint;

const updateAuthorization = {
	method: 'PUT',
	getUrl: ({authorizationKey}) => `/${API_VERSION}/authorizations/${authorizationKey}` as const,
} as const satisfies Endpoint<{authorizationKey: string}>;

const getAuthorization = {
	method: 'GET',
	getUrl: ({authorizationKey}) => `/${API_VERSION}/authorizations/${authorizationKey}` as const,
} as const satisfies Endpoint<{authorizationKey: string}>;

const deleteAuthorization = {
	method: 'DELETE',
	getUrl: ({authorizationKey}) => `/${API_VERSION}/authorizations/${authorizationKey}` as const,
} as const satisfies Endpoint<{authorizationKey: string}>;

const queryAuthorizations = {
	method: 'POST',
	getUrl: () => `/${API_VERSION}/authorizations/search` as const,
} as const satisfies Endpoint;

export {
	permissionTypeSchema,
	resourceTypeSchema,
	ownerTypeSchema,
	authorizationSchema,
	createAuthorizationRequestBodySchema,
	createAuthorizationResponseBodySchema,
	updateAuthorizationRequestBodySchema,
	getAuthorizationResponseBodySchema,
	queryAuthorizationsRequestBodySchema,
	queryAuthorizationsResponseBodySchema,
	createAuthorization,
	updateAuthorization,
	getAuthorization,
	deleteAuthorization,
	queryAuthorizations,
};

export type {
	PermissionType,
	ResourceType,
	OwnerType,
	Authorization,
	CreateAuthorizationRequestBody,
	CreateAuthorizationResponseBody,
	UpdateAuthorizationRequestBody,
	GetAuthorizationResponseBody,
	QueryAuthorizationsRequestBody,
	QueryAuthorizationsResponseBody,
};
